### PR TITLE
Implement SSH tunnels between client and podman server

### PR DESCRIPTION
**Pull request by jwhonce**
* client currently forks ssh client pending finding a well
  maintained ssh library for python.  Including support for
  AF_UNIX forwarding.

Signed-off-by: Jhon Honce <jhonce@redhat.com>

### CODE REVIEW

**Comment by jwhonce** [comment]
I'll remove the WIP label once I include more unit tests.

**Comment by rhatdan** [comment]
LGTM
What is the exact CLI that the user will use?
Can you add an environment variable to the system that controls indicates the remote server, so the user  does not need to specify it for every connection.

File: contrib/python/podman/client.py [hunk 1]
```diff
@@ -10,6 +10,119 @@
...
+                raise ValueError('username is required for remote_uri, format'
+                                 ' "ssh://user@hostname/path_to_socket".')
+            if remote.path == '':
+                raise ValueError('path is required for remote_uri, format'
+                                 ' "ssh://user@hostname/path_to_socket".')
+            if remote.hostname is None:
+                raise ValueError('hostname is required for remote_uri, format'
+                                 ' "ssh://user@hostname/path_to_socket".')
+
+            if kwargs.get('identity_file') is None:
```
Comment (TomSweeneyRedHat) [inline contrib/python/podman/client.py]: can we default this too if no provided?  ~/.ssh/id_rsa

File: contrib/python/podman/client.py [hunk 2]
```diff
@@ -20,37 +133,26 @@ class Client(object):
         >>> import podman
         >>> c = podman.Client()
         >>> c.system.versions
-    """
 
-    # TODO: Port to contextlib.AbstractContextManager once
-    #   Python >=3.6 required
+    Example remote podman:
```
Comment (TomSweeneyRedHat) [inline contrib/python/podman/client.py]: Don't you need three double qoutes (""") before this comment block?

File: contrib/python/podman/libs/tunnel.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        ]
+
+        if os.environ.get('PODMAN_DEBUG'):
+            cmd.append('-vvv')
+
+        self._tunnel = subprocess.Popen(cmd, close_fds=True)
+        for i in range(5):
+            if os.path.exists(self.context.local_socket):
+                break
+            time.sleep(1)
```
Comment (TomSweeneyRedHat) [inline contrib/python/podman/libs/tunnel.py]: nicely handled

**Comment by baude** [comment]
@rhatdan the connection profiles and such should be handled in the client itself, which I have somewhat started.  Right now, I'm on the fence on how that should be handled on the client side.  One could imagine a conf file for the client ... override capability in the command line and so forth.

File: contrib/python/podman/client.py [hunk 4]
```diff
@@ -10,6 +10,119 @@
...
+    def factory(cls,
+                uri=None,
+                interface='io.projectatomic.podman',
+                *args,
+                **kwargs):
+        """Construct a Client based on input."""
+        if uri is None:
+            raise ValueError('uri is required and cannot be None')
+        if interface is None:
+            raise ValueError('interface is required and cannot be None')
```
[the remark this answers is not in the document]
Comment (baude) [reply contrib/python/podman/client.py]: being this is a library, im ok with as-is.

File: contrib/python/podman/client.py (hunk 1 above)
Comment (baude) [reply contrib/python/podman/client.py]: if i use something like keychain, am i broken here?

**Comment by baude** [comment]
@rh-atomic-bot r+

**Comment by mheon** [comment]
@baude You also need to forward the terminal resize control FIFO, or have an API to send updates to it.

File: contrib/python/podman/client.py [hunk 5]
```diff
@@ -10,6 +10,119 @@
...
+            raise ValueError('path is required for uri, format'
+                             ' "unix://path_to_socket"')
+
+        if kwargs.get('remote_uri') or kwargs.get('identity_file'):
+            # Remote access requires the full tuple of information
+            if kwargs.get('remote_uri') is None:
+                raise ValueError('remote is required, format'
+                                 ' "ssh://user@hostname/path_to_socket".')
+            remote = urlparse(kwargs['remote_uri'])
+            if remote.username is None:
```
[the remark this answers is not in the document]
Comment (jwhonce) [reply contrib/python/podman/client.py]: @TomSweeneyRedHat @baude and I will add that to the CLI built on top of this work.

File: contrib/python/podman/client.py [hunk 6]
```diff
@@ -10,6 +10,119 @@
...
+                **kwargs):
+        """Construct a Client based on input."""
+        if uri is None:
+            raise ValueError('uri is required and cannot be None')
+        if interface is None:
+            raise ValueError('interface is required and cannot be None')
+
+        local_path = urlparse(uri).path
+        if local_path == '':
+            raise ValueError('path is required for uri, format'
```
[the remark this answers is not in the document]
Comment (jwhonce) [reply contrib/python/podman/client.py]: @TomSweeneyRedHat I added expected format to all messages.

File: contrib/python/podman/client.py (hunk 1 above)
Comment (jwhonce) [reply contrib/python/podman/client.py]: @baude Currently, that is not supported.  We'll have to research how to securely pull that information and not leave the password setting in the process table.

File: contrib/python/podman/client.py (hunk 2 above)
Comment (jwhonce) [reply contrib/python/podman/client.py]: @TomSweeneyRedHat The docstring is the whole block.  Both examples are included.

File: contrib/python/podman/libs/tunnel.py (hunk 3 above)
Comment (jwhonce) [reply contrib/python/podman/libs/tunnel.py]: @TomSweeneyRedHat @baude 's idea.  So far has worked well.  If the library ends up with more configuration items, these are prime candidates.

**Pull request merged**